bench: Add OrphanTxPool benchmark

hebasto opened the pull request:
This PR adds a new benchmark for adding/evicting orphan transactions to/from the `mapOrphanTransactions`.

This benchmark could be useful while considering relevant code changes (e.g., #19374) or changing the `DEFAULT_MAX_ORPHAN_TRANSACTIONS` parameter.

The output example:
```
$ ./src/bench/bench_bitcoin -filter="OrphanTxPool"
# Benchmark, evals, iterations, total, min, max, median
OrphanTxPool, 5, 10000, 2.3938, 4.76101e-05, 4.83722e-05, 4.77407e-0
```

hebasto committed:
bench: Add OrphanTxPool benchmark

jnewbery commented:
@hebasto can you provide motivation for this change? In practice, only one orphan transactions is ever evicted at a time (in `LimitOrphanTxSize()` which is only called after the 101st orphan has been added. Picking a random element from a 101-entry container is always going to be fast, whatever the underlying data structure, so I don't think there's any benefit to having a microbenchmark for this. On the other hand, there is a cost to adding it - it means that there's more work involved in changing or reviewing changes to the interface.

hebasto commented:
@jnewbery 
> @hebasto can you provide motivation for this change?

It was motivated by discussion with @sipa in #19374.

jnewbery commented:
I'm a -0.5 on this. I don't think it's worth adding a microbench for a different usage pattern than is used in the product, and for something that is always going to be quick.

MarcoFalke commented:
microbenchmarks are never going to match the exact usage pattern in the end product. This is true for all existing microbenchmarks (like the mempool eviction one, etc ...). Absent the details, I think having a benchmark is slightly preferable to not having a benchmark because it might catch accidental performance regressions.

And since we don't have any user-facing compatibility concerns for tests in general, adding and removing tests can be done much more easily than (let's say) an RPC method. In general I think that the cost of a pull request that adds a test that passes can only be negative or null (when it is testing nothing). And it should always be an uncontroversial option to remove a not-too-useful test that is in the way of other changes/refactoring/features.

I haven't looked at the specifics of this pull request, but I think in general we need more benchmarks (or fuzzers) for p2p tx relay, as that is the primary DoS attack vector on the live network.

jnewbery commented:
> In general I think that the cost of a pull request that adds a test that passes can only be negative or null (when it is testing nothing). And it should always be an uncontroversial option to remove a not-too-useful test that is in the way of other changes/refactoring/features.

I'm sorry - this is wrong. All code added to the repo has a cost - in maintenance, in review, in discussion as to whether to remove it later. Tests/benches should only be added if they're useful.

MarcoFalke commented on src/net_processing.cpp:
```diff
@@ -1018,6 +1019,13 @@ unsigned int LimitOrphanTxSize(unsigned int nMaxOrphans)
+    LOCK(g_cs_orphans);
+    SweepOutExpiredOrphans();
+    return EvictOrphanTxs(nMaxOrphans);
+}
```
Why is net_processing changed in a pull that adds a benchmark? Seems unnecessary to me

hebasto replied:
To benchmark `EvictOrphanTxs()` without sweeping out functionality of the current `LimitOrphanTxSize()`.

What could you suggest to change?

MarcoFalke replied:
If you explicitly want to disable the time based eviction, it could be done with mocktime in the benchmark, no?

hebasto replied:
[Updated](https://github.com/bitcoin/bitcoin/pull/19377#issuecomment-650013735).

MarcoFalke commented on src/bench/orphan_tx_pool.cpp:
```diff
@@ -0,0 +1,46 @@
+
+#include <vector>
+
+static constexpr CAmount CENT{1000000};
```
Why is this needed? values should not matter for performance

hebasto replied:
[Updated](https://github.com/bitcoin/bitcoin/pull/19377#issuecomment-650013735).

MarcoFalke commented on src/bench/orphan_tx_pool.cpp:
```diff
@@ -0,0 +1,46 @@
+
+static constexpr CAmount CENT{1000000};
+extern bool AddOrphanTx(const CTransactionRef& tx, NodeId peer);
+extern unsigned int EvictOrphanTxs(unsigned int nMaxOrphans);
```
Why is `extern` needed here?

hebasto replied:
[Updated](https://github.com/bitcoin/bitcoin/pull/19377#issuecomment-650013735).

maflcko reviewed (commented):
Concept ACK. I think adding a benchmark for orphan handling is useful and needed

hebasto commented:
Updated 059c8e6b31d55cad2ccf1dbb1cec0163cd58108d -> 2ee8c0b4be9e524f51572aba9ce4aafccbc0efef ([pr19377.01](https://github.com/hebasto/bitcoin/commits/pr19377.01) -> [pr19377.02](https://github.com/hebasto/bitcoin/commits/pr19377.02), [diff](https://github.com/hebasto/bitcoin/compare/pr19377.01..pr19377.02)):

- addressed all @MarcoFalke's comments

practicalswift commented:
Concept ACK

hebasto commented:
Rebased 2ee8c0b4be9e524f51572aba9ce4aafccbc0efef -> b0a576128669c7ebe61a53a3c3a9d1e21c4c844e ([pr19377.02](https://github.com/hebasto/bitcoin/commits/pr19377.02) -> [pr19377.03](https://github.com/hebasto/bitcoin/commits/pr19377.03)) due to the conflict with #18011.

jonatack commented on src/bench/orphan_tx_pool.cpp:
```diff
@@ -0,0 +1,42 @@
+    }
+
+    bench.batch(txs.size()).unit("byte").run([&] {
+        WITH_LOCK(g_cs_orphans, for (const auto& tx : txs) { AddOrphanTx(tx, 0); });
```
Concept ACK. Had a look as this "conflicts" with #22284, rebased it to current master and began building; it needs updating here:

```rake
bench/orphan_tx_pool.cpp:17:83: error: use of undeclared identifier 'g_cs_orphans'
bool AddOrphanTx(const CTransactionRef& tx, NodeId peer) EXCLUSIVE_LOCKS_REQUIRED(g_cs_orphans);
                                                                                  ^
bench/orphan_tx_pool.cpp:37:19: error: use of undeclared identifier 'g_cs_orphans'
        WITH_LOCK(g_cs_orphans, for (const auto& tx : txs) { AddOrphanTx(tx, 0); });
                  ^
bench/orphan_tx_pool.cpp:37:19: error: use of undeclared identifier 'g_cs_orphans'
3 errors generated.
```

jnewbery commented:
> Had a look as this "conflicts" with #22284, rebased it to current master and began building; it needs updating here and line 17 above:

This also conflicts with #21527, which is blocking quite a lot of important work in net_processing.

hebasto commented:
Marked this as a "draft" for now.
